designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/provider/ProcessingTypeDataProvider.scala

```diff
@@ -86,112 +90,119 @@ trait ProcessingTypeDataProvider[+Data, +CombinedData] {
 
   // TODO: We should return a generic type that can produce views for users with access rights to certain categories only.
   //       Thanks to that we will be sure that no sensitive data leak
-  final def combined: CombinedData = state.getCombined()
-
-  private[processingtype] def state: ProcessingTypeDataState[Data, CombinedData]
-
-  final def mapValues[TT](fun: Data => TT): ProcessingTypeDataProvider[TT, CombinedData] =
-    new TransformingProcessingTypeDataProvider[Data, CombinedData, TT, CombinedData](this, _.mapValues(fun))
-
-  final def mapCombined[CC](fun: CombinedData => CC): ProcessingTypeDataProvider[Data, CC] =
-    new TransformingProcessingTypeDataProvider[Data, CombinedData, Data, CC](this, _.mapCombined(fun))
+  final def combined: CombinedData = state.combinedDataTry.get
+
+  final protected[provider] def setStateValueAndNotifyObservers(
+      newValue: ProcessingTypeDataState[Data, CombinedData]
+  ): IO[Unit] = {
+    for {
+      _ <- stateRef.set(newValue)
+      _ <- observers.asScala
+        .map { observer =>
+          observer.notifyChange(newValue)
+        }
+        .toList
+        .sequence
+    } yield ()
+  }
 
-}
+  // Currently the whole application use this state synchronously, inside Future, because of that we unsafeRunSync()
+  // It is not so risky because we use separate IORuntime in ProcessingTypeDataProviders
+  // TODO: migrate application to IO
+  private def state: ProcessingTypeDataState[Data, CombinedData] = {
+    accessStateInCriticalSection(_.get).unsafeRunSync()
+  }
 
-private[processingtype] class TransformingProcessingTypeDataProvider[T, C, TT, CC](
-    observed: ProcessingTypeDataProvider[T, C],
-    transformState: ProcessingTypeDataState[T, C] => ProcessingTypeDataState[TT, CC]
-) extends ProcessingTypeDataProvider[TT, CC] {
+  final def mapValues[TT](fun: Data => TT): ProcessingTypeDataProvider[TT, CombinedData] = {
+    val childProvider =
+      new TransformingProcessingTypeDataProvider[Data, CombinedData, TT, CombinedData](this.state, _.mapValues(fun))
+    observers.add(childProvider)
+    childProvider
+  }
 
-  private val stateValue = new AtomicReference(transformState(observed.state))
+  final def mapCombined[CC](fun: CombinedData => CC): ProcessingTypeDataProvider[Data, CC] = {
+    val childProvider =
+      new TransformingProcessingTypeDataProvider[Data, CombinedData, Data, CC](this.state, _.mapCombined(fun))
+    observers.add(childProvider)
+    childProvider
+  }
 
-  override private[processingtype] def state: ProcessingTypeDataState[TT, CC] = {
-    stateValue.updateAndGet { currentValue =>
-      val currentObservedState = observed.state
-      if (currentObservedState.stateIdentity != currentValue.stateIdentity) {
-        transformState(currentObservedState)
-      } else {
-        currentValue
+  def close(): IO[Unit] = {
```

Comment:
In my opinion, we shouldn't allow this method to be overridden (wrong child implementation may break this base class's internals).